modules/kmm-customizing-upgrades-for-kernel-modules.adoc

```diff
@@ -0,0 +1,54 @@
+// Module included in the following assemblies:
+//
+// * hardware_enablement/kmm-kernel-module-management.adoc
+
+:_content-type: PROCEDURE
+[id="kmm-customizing-upgrades-for-kernel-modules_{context}"]
+= Customizing upgrades for kernel modules
+
+Use this procedure to upgrade the kernel module while running maintenance operations on the node, including rebooting the node, if needed. To minimize the impact on the workloads running in the cluster, run the kernel upgrade process sequentially, one node at a time.
+
+[NOTE]
+====
+This procedure requires knowledge of the workload utilizing the kernel module and must be managed by the cluster administrator.
+====
+
+.Prerequisites
+
+* Terminate all user application workloads on the node or move them to another node.
+
+* Unload the currently loaded kernel module.
+
+* Ensure that the user workload (the application running in the cluster that is accessing kernel module) is not running on the node prior to kernel module unloading and that the workload is back running on the node after the new kernel module version has been loaded.
+
+* Ensure that the device plugin managed by KMM on the node is unloaded before the upgrade process and reloaded following the upgrade process.
+
+* Due to Kubernetes limitations in label names, the combined length of `Module` name and namespace must not exceed 39 characters.
+
+
+.Procedure
+
+. Before upgrading, set the `kmm.node.kubernetes.io/version-module.<module-namespace>.<module-name>=$moduleVersion` label on all the nodes that will be used by the kernel module.
+
+. Update the following fields in the `Module` CR:
+- `containerImage` (to the appropriate kernel version)
+- `version`
++
+The update should be atomic; that is, both the `containerImage` and `version` fields must be updated simultaneously.
+
+. Terminate any workload using the kernel module on the node being upgraded.
+
+. Remove the `kmm.node.kubernetes.io/version-module.<module-namespace>.<module-name>` label on the node. This unloads the kernel module from the node.
+
+. If required, perform any additional maintenance required on the node for the kernel module upgrade.
+
+. Add the `kmm.node.kubernetes.io/version-module.<module-namespace>.<module-name>=$moduleVersion` label to the node. `$moduleVersion` should be equal to the new value of the `version` field in the `Module` CR.
+
+. Restore any workload that leverages the kernel module on the node.
+
+[NOTE]
+====
+If no additional upgrading (Step 4) is needed, then the process is simplified to changing the `kmm.node.kubernetes.io/version-module.<module-namespace>.<module-name>=$moduleVersion` label. Steps 3, 4 and 5 are thus unified into a single step:
```

Comment:
I still feel this note is confusing tbh.

Reply:
@Aidan I spoke with Costi and we came up this - "If no additional upgrading is needed, you can skip Steps 3 through 6 by updating the `[kmm.node.kubernetes.io/version-module](http://kmm.node.kubernetes.io/version-module).<module-namespace>.<module-name>` label value to the new `$moduleVersion` as set in the `Module`."

Reply:
OK cool. I presume what we are getting at here is that you can only upgrade a single kernel module at a time. Is this the case?

Reply:
@aireilly From Costi - "you can do multiple at the same time, if you respect the procedure"

Reply:
/label merge-review-needed